runtime/vm/xz64/UpcallThunkGen.cpp

```diff
@@ -0,0 +1,767 @@
+	I_32 roundedCodeSize = 0;
+	I_8 *thunkMem = NULL;
+
+	// stack space to store arguments or return values,
+	frameSize = offsetToParamArea + numStackSlots * 8;
+
+	// 8 byte align the thunk memory.
+	I_32 extraInstructionsSize = 64; // Needed for backing up and restoring data before/after we invoke the common dispatcher
+	roundedCodeSize = ((( numInstructions * 6 + extraInstructionsSize) / 8) + 1) * 8;
+	metaData->thunkSize = roundedCodeSize + 8;
+	thunkMem = (I_8 *)vmFuncs->allocateUpcallThunkMemory(metaData);
+	if (NULL == thunkMem) {
+		return NULL;
+	}
+	metaData->thunkAddress = (void *)thunkMem;
+
+	//Generate the instruction sequence to copy data from C side to Java side.
+	I_32 stackSlotIndex = 0;
+	I_32 gprIndex = 2;
+	I_32 fprIndex = 0;
+	I_32 C_SP_RegisterNum = 15;
+	// save the original return address into memory
+	thunkMem += STG(thunkMem, 14, 0, C_SP_RegisterNum, 112);
+	// Store the current C stack pointer at the first 8-byte slot of the new frame.
+	thunkMem += STG(thunkMem, C_SP_RegisterNum, 0, C_SP_RegisterNum, -frameSize);
+	// Then update R15 with the new c stack pointer.
+	thunkMem += AGFI(thunkMem, C_SP_RegisterNum, -frameSize);
+
+	if (hiddenParameter) {
+		// if there's a hidden parameter, then store it.
+		thunkMem += STG(thunkMem, gprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
+		gprIndex++;
+		stackSlotIndex += 1;
+	}
+	
+	// Now loop through the arguments and generate instructions to store arguments that are in registers into the callee frame.
+	// Arguments that are already in the caller frame don't need to be stored again.
+	for (I_32 i = 0; i < lastSigIdx; i++)
+	{
+		switch(sigArray[i].type) {
+			case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+			case J9_FFI_UPCALL_SIG_TYPE_INT64:
+				// Only store the value onto the stack if it's not already in the caller's frame (i.e. in a register).
+				if (gprIndex <= 6) {
+					thunkMem += STG(thunkMem, gprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8)); 
+				}	
+				gprIndex += 1;
+				stackSlotIndex += 1;
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+				if (fprIndex <= 6) {
+					thunkMem += STEY(thunkMem, fprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
+				}
+				fprIndex += 2;
+				stackSlotIndex += 1;
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+				if (fprIndex <= 6) {
+					thunkMem += STDY(thunkMem, fprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
+				}
+				fprIndex += 2;
+				stackSlotIndex += 1;
+				break;
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP:
+				if (sigArray[i].sizeInByte == 4) {
+					if (fprIndex <= 6) {
+						thunkMem += STEY(thunkMem, fprIndex, 0, C_SP_RegisterNum, offsetToParamArea + (stackSlotIndex * 8));
```

Comment:
this looks wrong:  since single-precision FP is represented in the left part of the register, you cannot jump over structOffset to return the address later on.

Reply:
The structOffset will be zero when the struct size is 1, 2, 4 or 8 bytes.

On zLinux only structs of size 1, 2, 4, and 8 can be passed by value in a register. This is usually done in a GPR. 

The only scenario where this is done in an FPR is if we have a pure float/double struct with a single element (i.e. structs of size 4 and 8 bytes).

So I don't think we will end up in a problematic situation. Unless I'm missing sometime else here?

Reply:
No. the point is you cannot handle struct-of-1/2/4 bytes the same way with FPR and GPR arguments. the reason is: FPR is only populated in the left and GPR is only populated on the right. Imagine a single pure float struct to be passed ... it is passed by value in FPR with left half register populated ... you store it back with STEY at offset 0 in the stack slot ... later on, you have a structOffset value of 4 ... the return address actually points at the right half of the slot.

Reply:
This has been fixed now. `structOffset` will be zero for FPR argument types (including structs passed by value in FPR).